infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -0,0 +1,133 @@
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.google.common.annotations.VisibleForTesting;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.view.Address;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Sequencer server cache.
+ * Contains transaction conflict-resolution data structures.
+ */
+@Slf4j
+public class SequencerServerCache {
+    /**
+     * TX conflict-resolution information:
+     * <p>
+     * a cache of recent conflict keys and their latest global-log position.
+     */
+    private final Cache<String, Long> conflictToGlobalTailCache;
+
+    /**
+     * A "wildcard" representing the maximal update timestamp of
+     * all the conflict keys which were evicted from the cache
+     */
+    @Getter
+    private long maxConflictWildcard = Address.NOT_FOUND;
+
+    /**
+     * maxConflictNewSequencer represents the max update timestamp of all the conflict keys
+     * which were evicted from the cache by the time this server is elected
+     * the primary sequencer. This means that any snapshot timestamp below this
+     * actual threshold would abort due to NEW_SEQUENCER cause.
+     */
+    @Getter
+    private long maxConflictNewSequencer = Address.NOT_FOUND;
+
+    @VisibleForTesting
+    public SequencerServerCache(long cacheSize, CacheWriter<String, Long> writer) {
+        this.conflictToGlobalTailCache = Caffeine.newBuilder()
+                .maximumSize(cacheSize)
+                .writer(writer)
+                .executor(Runnable::run)
+                .recordStats()
+                .build();
+    }
+
+    /**
+     * Limited cache by size.
+     * Eviction policy described here https://github.com/ben-manes/caffeine/wiki/Writer
+     * @param cacheSize cache size
+     */
+    public SequencerServerCache(long cacheSize) {
+        CacheWriter<String, Long> writer = new CacheWriter<String, Long>() {
+            @Override
+            public void write(@Nonnull String key, @Nonnull Long value) {
+                //ignore
+            }
+
+            @Override
+            public void delete(@Nonnull String key, @Nullable Long globalAddress, @Nonnull RemovalCause cause) {
+                if (cause == RemovalCause.REPLACED) {
+                    return;
+                }
+
+                log.trace(
+                        "Updating maxConflictWildcard. Old = '{}', new ='{}' conflictParam = '{}'. Cause = '{}'",
+                        maxConflictWildcard, globalAddress, key, cause
+                );
+
+                if (globalAddress == null) {
+                    globalAddress = Address.NOT_FOUND;
+                }
+                maxConflictWildcard = Math.max(globalAddress, maxConflictWildcard);
+            }
+        };
+
+        this.conflictToGlobalTailCache = Caffeine.newBuilder()
+                .maximumSize(cacheSize)
+                .executor(Runnable::run)
```

Comment:
Can you add a comment about how this "trick" makes the cache use the accessing thread and not a background thread.

Reply:
done

> Slava, please follow the code guidelines! The line break is at 100 not 120.
> 
> Also, you need a lot more documentation, especially WRT concurrency. I assume the secret ingredient is ".executor(Runnable::run)".

Line length 120 - https://github.com/CorfuDB/CorfuDB/blob/5540d594eba3b7adfed1eb9eba505ce90cc8be40/checks.xml#L88

P.S. Will add some doc